addons/sale_margin/models/sale_order.py

```diff
@@ -46,7 +46,7 @@ def create(self, vals):
 
         # Calculation of the margin for programmatic creation of a SO line. It is therefore not
         # necessary to call product_id_change_margin manually
-        if 'purchase_price' not in vals:
+        if 'purchase_price' not in vals and 'display_type' not in vals:
```

Comment:
I would go for:
`if 'purchase_price' not in vals and ('display_type' not in vals or vals['display_type'] is False):`

Reply:
You are right, I misread the super:
https://github.com/odoo/odoo/blob/12cca6a5fdf82f24ba536efab465533f6066fb47/addons/sale/models/sale.py#L1031-L1032
It's safer this way, and this more resilient to other changes in sale. 
Thanks for the catch.

Reply:
or:
`if 'purchase_price' not in vals and ('display_type' not in vals or not vals['display_type'])`

Reply:
or
``` python
if 'purchase_price' not in vals and not vals.get('display_type')
```
if you want to spare a few chars

Reply:
but i see that robodoo is already on it so whatever :p